### ISSUE TITLE
Expose same-thread executor

### ISSUE THREAD

**Issue by Diggsey** [alexcrichton/futures-rs#322]
Currently it's possible to execute a future synchronously via `wait()`, which relies on an internal implementation of an executor, polling the future on the same thread until it completes.

I think the library should expose this executor, and allow multiple futures to be spawned onto it.

eg.
```
let exec = executor::blocking();
exec.spawn(future_a);
exec.spawn(future_b);
let result = exec.wait(future_c);
```

All futures will be polled until `future_c` completes. Only the result from `future_c` will be returned. The `wait()` method on futures will become shorthand for `executor::blocking().wait(self)`.

**Comment by alexcrichton** [comment]
The "executor" behind `wait` is *very* different than what you're thinking of hear. This can, I believe, be built entirely out of tree (and probably should!). The fundamental building block is just `task::spawn`, and after that how to poll a future is entirely up to you. Something like this report seems like it would involve local caches of futures, perhaps unpark events, etc. (none of which is used in `Future::wait`)

**Comment by bmeck** [comment]
I had the same problem trying to figure out how to notify my spawn that I need the future to be polled for streams that work on same thread https://gist.github.com/bmeck/761783e8155f7f623062d45c254e27a6 . Parking the thread immediately prevents the Stream from telling any spawn on same thread from polling/preventing the park.

Perhaps an example would be prudent as this also is not in the tutorial and seems rather important to understanding how to use the very mysterious `Executor`:

http://alexcrichton.com/futures-rs/futures/stream/trait.Stream.html#tymethod.poll

> then implementations will ensure that the current task will be notified when the next value may be ready.

Implementations of (Executors?)? *How* do I tell the currently used one for a stream should be notified? How do I find out the current thing (CURRENT_TASK is not public)?

http://alexcrichton.com/futures-rs/futures/stream/trait.Stream.html#tymethod.poll

> The current task will be automatically scheduled for notification when the sink may be ready to receive new values.

How does it test this? When .send or .poll_complete is fired it checks this somehow? What is is being scheduled on/against (an Executor)?

**Comment by alexcrichton** [comment]
@bmeck the documentation on `poll` isn't related to executors and spawns and such. That documentation is referencing implementations of the trait being described (in this case `Stream`). Blocking and notifying a task is done through `task::park` and `Task::unpark`.

**Comment by bmeck** [comment]
@alexcrichton there is no connection to task::park / unpark if you can't get a hold of the current one?

**Comment by bmeck** [comment]
@alexcrichton to clarify further: even with park/unpark , if you have a `Sink + Stream` as in my example, you don't have a clean way to generate the `unpark` to my knowledge. You have a workflow like:

```
let (sink,stream) = new_duplex();
let exec_stream = spawn(stream);
let exec_stream = spawn(sink);
// at this point stream needs to get a hold of `unpark`
let unpark_sink = `???`;
stream.ondrain(unpark_sink); // unpark when stream drains
sink.ondrain(|| {
  // request more data for the sink?
});
exec_stream.wait(); // wait until output finishes
````

I am unclear on how to actually generate these bits given the docs. It appears `spawn(stream)` would need to grab the `unpark` ready on the first `poll_complete`/`poll` maybe?

**Comment by alexcrichton** [comment]
@bmeck I'm pretty confused by your example unfortunately, I'm not sure what's going on :(

In general you should never have to worry about park/unpark unless you're writing an implementation of `Future`. You should never have to worry about `Arc<Unpark>` or such unless you're writing your own custom executor. And in general you shouldn't have to worry about an executor at all unless you're explicitly trying to schedule futures in particular locations.

**Comment by bmeck** [comment]
@alexcrichton I would agree on being confused on both ends.

My example is implementing a `ChunkedEncodingParser` ala (https://en.wikipedia.org/wiki/Chunked_transfer_encoding). It is:

* A `Sink` that consumes `u8` and places any byte that needs to be output in the single `buffer`.
* A `Stream` that produces `u8` when it `buffer` is filled.
* A `Future` for when the parser is unable to continue.

So yes, I am implementing a Future/Sink/Stream. All of which appear to need to use park/unpark given the previous comments here, which goes to strange lands when reading the docs. They mention a need to do something (notify the current task), but lack an explanation of how to do so.

After many attempts at this, I do not believe you can achieve single threaded duplex piping without park/unpark being done by the Sink/Stream/Future. Sadly, I cannot figure out any mixture of API calls that has meaningful effect on the executor after trying to read/understand the docs.

**Comment by alexcrichton** [comment]
@bmeck want to stop by the tokio gitter channel? I think it may be best to take this conversation there instead of continuing on this issue.

**Comment by alexcrichton** [comment]
FWIW this [exists in the test suite](https://github.com/alexcrichton/futures-rs/blob/master/tests/support/local_executor.rs) and I'd be totally fine if someone wanted to extract this to a crate!

(could also add timeouts and such externally)

**Comment by carllerche** [comment]
This should be handled by `CurrentThread` which exists on the tokio-reform branch.

Relates to #645

**Comment by carllerche** [comment]
I'm going to close this issue as it will be resolved by #645.
